lib/ansible/plugins/inventory/netbox.py

```diff
@@ -0,0 +1,352 @@
+        regions = self.get_resource_list(api_url=url)
+        self.regions_lookup = dict((region["id"], region["name"]) for region in regions)
+
+    def refresh_tenants_lookup(self):
+        url = urljoin(self.api_endpoint, "/api/tenancy/tenants/?limit=0")
+        tenants = self.get_resource_list(api_url=url)
+        self.tenants_lookup = dict((tenant["id"], tenant["name"]) for tenant in tenants)
+
+    def refresh_racks_lookup(self):
+        url = urljoin(self.api_endpoint, "/api/dcim/racks/?limit=0")
+        racks = self.get_resource_list(api_url=url)
+        self.racks_lookup = dict((rack["id"], rack["name"]) for rack in racks)
+
+    def refresh_device_roles_lookup(self):
+        url = urljoin(self.api_endpoint, "/api/dcim/device-roles/?limit=0")
+        device_roles = self.get_resource_list(api_url=url)
+        self.device_roles_lookup = dict((device_role["id"], device_role["name"]) for device_role in device_roles)
+
+    def refresh_device_types_lookup(self):
+        url = urljoin(self.api_endpoint, "/api/dcim/device-types/?limit=0")
+        device_types = self.get_resource_list(api_url=url)
+        self.device_types_lookup = dict((device_type["id"], device_type["model"]) for device_type in device_types)
+
+    def refresh_manufacturers_lookup(self):
+        url = urljoin(self.api_endpoint, "/api/dcim/manufacturers/?limit=0")
+        manufacturers = self.get_resource_list(api_url=url)
+        self.manufacturers_lookup = dict((manufacturer["id"], manufacturer["name"]) for manufacturer in manufacturers)
+
+    def refresh_lookups(self):
+        self.refresh_sites_lookup()
+        self.refresh_regions_lookup()
+        self.refresh_tenants_lookup()
+        self.refresh_racks_lookup()
+        self.refresh_device_roles_lookup()
+        self.refresh_device_types_lookup()
+        self.refresh_manufacturers_lookup()
+
+    def validate_query_parameters(self, x):
+        if not (isinstance(x, dict) and len(x) == 1):
+            self.display.warning("Warning query parameters %s not a dict with a single key." % str(x))
+            return
+
+        k = x.keys()[0]
+        v = x.values()[0]
+
+        if not (k in ALLOWED_DEVICE_QUERY_PARAMETERS or k.startswith("cf_")):
+            self.display.warning("Warning: %s not in %s or starting with cf (Custom field)" % (k, ALLOWED_DEVICE_QUERY_PARAMETERS))
+            return
+        return k, v
+
+    def refresh_url(self):
+        query_parameters = [("limit", 0)]
+        query_parameters.extend([self.validate_query_parameters(x)
+                                 for x in self.query_filters
+                                 if self.validate_query_parameters(x)])
+        self.device_url = self.api_endpoint + "/api/dcim/devices/" + "?" + urlencode(query_parameters)
+
+    def fetch_hosts(self):
+        return self.get_resource_list(self.device_url)
+
+    def extract_name(self, host):
+        return host["name"] or str(uuid.uuid4())
```

Comment:
@pilou- Is it possible that if the function is called several times, we will get a different UUID? Should we assign this UUID to the host dictionary?

Reply:
I did think about that yesterday, but since that's called only once, that should be ok. But it could be a good idea to handle the case, in case the code start to do use extract_name more than once.

Reply:
Indeed, `extract_name` needs to be called only once per host. In order to do that add a parameter to `add_host_to_groups` method:
```
self.add_host_to_groups(host, name)
```

```
def add_host_to_groups(self, host, name):
    ...
```